server/queue/queue-memory/src/main/java/org/apache/james/queue/memory/MemoryMailQueueFactory.java

```diff
@@ -241,5 +253,15 @@ public void done(boolean success) throws MailQueue.MailQueueException {
                 queue.enQueue(mail);
             }
         }
+
+        @Override
+        public long getDelay(TimeUnit unit) {
+            return LocalDateTime.now().until(delivery, Temporals.chronoUnit(unit));
+        }
+
+        @Override
+        public int compareTo(Delayed o) {
+            return Math.toIntExact(getDelay(TimeUnit.MILLISECONDS) - o.getDelay(TimeUnit.MILLISECONDS));
```

Comment:
you don't compare on other fields?

Reply:
it's Comparable<Delayed> so there's only this field I guess

Reply:
> The quotient for this total order is:

       {(x, y) such that x.compareTo(y) == 0}.
 

> It follows immediately from the contract for compareTo that the quotient is an equivalence relation on C, and that the natural ordering is a total order on C. When we say that a class's natural ordering is consistent with equals, we mean that the quotient for the natural ordering is the equivalence relation defined by the class's equals(Object) method:

     {(x, y) such that x.equals(y)}. 

I understand that in case of 0, compare should be equivalent to equals.

Reply:
Ok, so my comment got truncated because of generics.
I meant, it's `Comparable<Delayed>` so you just have to compare on Delayed fields

Reply:
oh, I see. Strange but let's go with this.